### PR TITLE
Optimize time and value in filter by TsFile statistics.

### PR DESCRIPTION
**Pull request by Sh-Zh-7**
As title.

### CODE REVIEW

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 1]
```diff
@@ -765,7 +771,89 @@ public ValueIn(ByteBuffer buffer) {
...
+        return true;
+      }
+
+      if (statistics.isPresent()) {
+        T valuesMin = (T) statistics.get().getMinValue();
+        T valuesMax = (T) statistics.get().getMaxValue();
+        // All values are same
+        if (valuesMin == valuesMax) {
+          for (T val : candidates) {
+            if (valuesMin == val) {
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: why not use candidates.contains(valuesMin) ?

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 2]
```diff
@@ -765,7 +771,89 @@ public ValueIn(ByteBuffer buffer) {
...
+    }
+
+    @Override
+    public boolean allSatisfy(IMetadata metadata) {
+      Optional<Statistics<? extends Serializable>> statistics =
+          metadata.getMeasurementStatistics(measurementIndex);
+
+      // All values are null, and candidate is one null
+      if ((!statistics.isPresent() || isAllNulls(statistics.get()))
+          && candidates.size() == 1
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: no need to be size == 1 ?

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 3]
```diff
@@ -765,7 +771,89 @@ public ValueIn(ByteBuffer buffer) {
...
+        return true;
+      }
+
+      // All values are same
+      if (statistics.isPresent()) {
+        T valuesMin = (T) statistics.get().getMinValue();
+        T valuesMax = (T) statistics.get().getMaxValue();
+        // All values are same
+        if (valuesMin == valuesMax) {
+          for (T val : candidates) {
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: why not use contains?

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 4]
```diff
@@ -765,7 +771,89 @@ public ValueIn(ByteBuffer buffer) {
...
+            if (valuesMin != val) {
+              return false;
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: wrong?

Any values in set equals to valuesMin, should return true

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 5]
```diff
@@ -794,6 +886,16 @@ public boolean valueSatisfy(Object value) {
       return !candidates.contains(value);
     }
 
+    @Override
+    protected boolean canSkip(Statistics<? extends Serializable> statistics) {
+      throw new NotImplementedException();
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: return false? why throw exception here?

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 6]
```diff
@@ -794,6 +886,16 @@ public boolean valueSatisfy(Object value) {
...
+    }
+
+    @Override
+    protected boolean allSatisfy(Statistics<? extends Serializable> statistics) {
+      throw new NotImplementedException();
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: return false? why throw exception here?

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
Add UT For all you changed filters.

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 1 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Fixed.

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 2 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Nice catch, fixed.

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 3 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Affected by ORC. 😅

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 4 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Fixed.

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 5 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Would IoTDB push down NOT IN operator for now?

Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Anyway, return false won't be wrong. Fixed.

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java (hunk 6 above)
Comment (Sh-Zh-7) [reply java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: Ditto.

**Comment by Sh-Zh-7** [comment]
PTAL~

File: java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java [hunk 7]
```diff
@@ -765,7 +771,78 @@ public ValueIn(ByteBuffer buffer) {
...
+      // All values are not null, but candidate is one null value
+      if (!metadata.hasNullValue(measurementIndex) && candidates.size() == 0) {
+        return true;
+      }
+
+      if (statistics.isPresent()) {
+        T valuesMin = (T) statistics.get().getMinValue();
+        T valuesMax = (T) statistics.get().getMaxValue();
+        // All values are same
+        if (valuesMin == valuesMax) {
```
Comment (JackieTien97) [inline java/tsfile/src/main/java/org/apache/tsfile/read/filter/operator/ValueFilterOperators.java]: valuesMin.compareTo(valuesMax) == 0?

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**